File: frontend/src/api/copilot.ts
/**
 * Typed data layer for the panel copilot (CLAUDE.md rule 9, uiux-spec §17):
 * "Ask AI to analyze" on failed jobs and the natural-language command palette.
 * Everything rides through apiClient — the backend owns the sanitisation, the
 * model routing, and (for palette proposals) the exact run.method/path/body, so
 * the frontend never builds a command or shell string of its own.
 */

import { apiClient } from './client'

export type AnalysisStatus = 'pending' | 'running' | 'success' | 'failure'

export interface JobAnalysis {
  id: number
  job_id: number
  status: AnalysisStatus
  model: string | null
  root_cause: string | null
  suggested_fix: string | null
  summary: string | null
  error: string | null
  requested_by: number | null
  created_at: string
  completed_at: string | null
}

/** One NL-palette proposal. `run` is the server-authored call to execute on confirm. */
export interface NLProposal {
  title: string
  action_name: string
  summary: string
  site_id: number
  site_name: string
  params: Record<string, string>
  allowed: boolean
  confirm: boolean
  run: {
    method: string
    path: string
    body: Record<string, unknown>
  }
}

export interface NLResolveResponse {
  resolved: boolean
  intent: string | null
  reason: string | null
  proposals: NLProposal[]
}

export const copilotApi = {
  analyzeJob: (jobId: number) => apiClient.post<JobAnalysis>(`/api/jobs/${jobId}/analyze`),
  getAnalysis: (jobId: number) => apiClient.get<JobAnalysis>(`/api/jobs/${jobId}/analyze`),
  resolvePalette: (q: string) => apiClient.post<NLResolveResponse>('/api/palette/resolve', { q }),
}

File: backend/app/schemas/copilot.py
"""Request/response schemas for the panel copilot (session 5.2)."""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field

from app.models.job_analysis import JobAnalysis


class JobAnalysisOut(BaseModel):
    """One AI analysis of a failed job. Rendered by the "Ask AI to analyze"
    panel; polled until `status` is terminal."""

    id: int
    job_id: int
    status: str
    model: str | None = None
    root_cause: str | None = None
    suggested_fix: str | None = None
    summary: str | None = None
    error: str | None = None
    requested_by: int | None = None
    created_at: datetime
    completed_at: datetime | None = None

    @classmethod
    def from_model(cls, a: JobAnalysis) -> JobAnalysisOut:
        return cls(
            id=a.id,
            job_id=a.job_id,
            status=a.status,
            model=a.model,
            root_cause=a.root_cause,
            suggested_fix=a.suggested_fix,
            summary=a.summary,
            error=a.error,
            requested_by=a.requested_by,
            created_at=a.created_at,
            completed_at=a.completed_at,
        )


class NLResolveRequest(BaseModel):
    q: str = Field(default="", max_length=400)


class NLProposalOut(BaseModel):
    """A confirm-before-run action mapped to an existing registered template."""

    title: str
    action_name: str
    summary: str
    site_id: int
    site_name: str
    params: dict
    allowed: bool
    confirm: bool
    run: dict


class NLResolveResponse(BaseModel):
    resolved: bool
    intent: str | None = None
    reason: str | None = None
    proposals: list[NLProposalOut] = []
